
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>About pygtftk &#8212; gtftk 0.9.6 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="author" title="About these documents" href="#" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installing pygtftk/gtftk" href="installation.html" />
    <link rel="prev" title="&lt;no title&gt;" href="index.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installing pygtftk/gtftk"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="&lt;no title&gt;"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 0.9.6 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="about-pygtftk">
<h1>About pygtftk<a class="headerlink" href="#about-pygtftk" title="Permalink to this headline">¶</a></h1>
<p>The <strong>Python GTF toolkit (pygtftk) package</strong> is intented to ease handling of GTF files (Gene Transfer Format). The pygtftk package is compatible with Python 2.7 and relies on <strong>libgtftk</strong>, a library of functions <strong>written in C</strong>.</p>
<p>The package comes with a set of <strong>UNIX commands</strong> that can be accessed through the <strong>gtftk main Unix program</strong>. The gtftk program expose several subcommands than can be piped, for instance, to filter, convert, extract or delete data from GTF files. The gtftk set of Unix commands, can be easily extended using a basic plugin architecture. All these aspects are covered in the help section.</p>
<p>While the gtftk Unix program comes with hundreds of unitary and functional tests, it is still upon  active thus feel free to post any problem or required enhancement through the github interface.</p>
<div class="section" id="warning-about-supported-gtf-file-formats">
<h2>warning about supported GTF file formats<a class="headerlink" href="#warning-about-supported-gtf-file-formats" title="Permalink to this headline">¶</a></h2>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Most of the commands of the gtftk suite are designed to handle files in <strong>ensembl</strong> GTF format and thus require <strong>transcript and gene features/lines</strong> in the GTF. All lines must contain a transcript_id and gene_id value except the <strong>gene feature</strong> that should contain only the gene_id (<strong>see get_example command for an example</strong>). Transcript and gene lines will be used when required to get access to transcript and gene coordinates. This solution was choosen to define a reference GTF file format for gtftk (since Ensembl format is probably the most widely used).</p>
</div>
<p>You can use the <strong>convert_ensembl</strong> subcommand to convert your non- (or old) ensembl format to current ensembl format.</p>
<p>Below an example in which we first select only exon features then use <strong>convert_ensembl</strong> to re-generate gene and transcript features using <strong>convert_ensembl</strong> .</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example | gtftk select_by_key -k feature  -v exon | head -n 10
chr1	gtftk	exon	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;G0001T002&quot;; exon_id &quot;G0001T002E001&quot;;
chr1	gtftk	exon	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;G0001T001&quot;; exon_id &quot;G0001T001E001&quot;;
chr1	gtftk	exon	180	189	.	+	.	gene_id &quot;G0002&quot;; transcript_id &quot;G0002T001&quot;; exon_id &quot;G0002T001E001&quot;;
chr1	gtftk	exon	50	54	.	-	.	gene_id &quot;G0003&quot;; transcript_id &quot;G0003T001&quot;; exon_id &quot;G0003T001E001&quot;;
chr1	gtftk	exon	57	61	.	-	.	gene_id &quot;G0003&quot;; transcript_id &quot;G0003T001&quot;; exon_id &quot;G0003T001E002&quot;;
chr1	gtftk	exon	65	68	.	+	.	gene_id &quot;G0004&quot;; transcript_id &quot;G0004T002&quot;; exon_id &quot;G0004T002E001&quot;;
chr1	gtftk	exon	71	71	.	+	.	gene_id &quot;G0004&quot;; transcript_id &quot;G0004T002&quot;; exon_id &quot;G0004T002E002&quot;;
chr1	gtftk	exon	74	76	.	+	.	gene_id &quot;G0004&quot;; transcript_id &quot;G0004T002&quot;; exon_id &quot;G0004T002E003&quot;;
chr1	gtftk	exon	65	68	.	+	.	gene_id &quot;G0004&quot;; transcript_id &quot;G0004T001&quot;; exon_id &quot;G0004T001E001&quot;;
chr1	gtftk	exon	71	71	.	+	.	gene_id &quot;G0004&quot;; transcript_id &quot;G0004T001&quot;; exon_id &quot;G0004T001E002&quot;;
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example | gtftk select_by_key -k feature  -v exon | gtftk  convert_ensembl | head -n 10
chr1	gtftk	gene	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	transcript	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;G0001T002&quot;;
chr1	gtftk	exon	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;G0001T002&quot;; exon_id &quot;G0001T002E001&quot;;
chr1	gtftk	transcript	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;G0001T001&quot;;
chr1	gtftk	exon	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;G0001T001&quot;; exon_id &quot;G0001T001E001&quot;;
chr1	gtftk	gene	180	189	.	+	.	gene_id &quot;G0002&quot;;
chr1	gtftk	transcript	180	189	.	+	.	gene_id &quot;G0002&quot;; transcript_id &quot;G0002T001&quot;;
chr1	gtftk	exon	180	189	.	+	.	gene_id &quot;G0002&quot;; transcript_id &quot;G0002T001&quot;; exon_id &quot;G0002T001E001&quot;;
chr1	gtftk	gene	50	61	.	-	.	gene_id &quot;G0003&quot;;
chr1	gtftk	transcript	50	61	.	-	.	gene_id &quot;G0003&quot;; transcript_id &quot;G0003T001&quot;;
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk convert_ensembl -h
  Usage: gtftk convert_ensembl [-i GTF] [-o GTF] [-n] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Convert the GTF file to ensembl format. It will essentially add a &#39;transcript&#39; feature and &#39;gene&#39;
     feature when required. This command can be viewed as a &#39;groomer&#39; command for those starting
     with a non ensembl GTF.

  Notes:
     *  The gtftk program is designed to handle files in ensembl GTF format. This means that the GTF
     file provided to gtftk must contain transcript and gene feature/lines. They will be used to get
     access to transcript and gene coordinates whenever needed. This solution was choosen to define
     a reference GTF file format for gtftk (since Ensembl format is probably the most widely used).
     *  Almost all commands of gtftk use transcript_id or gene_id as keys to perform operation on
     genomic coordinates. One of the most common issue when working with  gene coordinates is the
     lack  of non ambiguous gene or transcript names For instance, a refSeq sequence ID used as
     transcript_id can be associated to  several chromosomal locations as a sequence may be
     duplicated. These identifiers are ambiguous and thus should be avoid. Use UCSC or ensembl IDs
     instead.

  Version:  2018-02-06

Arguments:
 -i, --inputfile          Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile         Output file. (default: &lt;stdout&gt;)
 -n, --no-check-gene-chr  By default the command raise an error if several chromosomes are associated with the same gene_id. Disable this behaviour (but you should better think about what it means...). (default: False)

Command-wise optional arguments:
 -h, --help               Show this help message and exit.
 -V, --verbosity          Increase output verbosity. (default: 0)
 -D, --no-date            Do not add date to output file names. (default: False)
 -C, --add-chr            Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir            Keep all temporary files into this folder. (default: None)
 -A, --keep-all           Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file        Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">any comment line (<em>i.e.</em> starting with #) or empty line in the gtf file will be ignore (discarded) by gtftk.</p>
</div>
</div>
<div class="section" id="naming-conventions">
<h2>Naming conventions<a class="headerlink" href="#naming-conventions" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">We will use the terms <strong>attribute or key</strong> for any descriptor found in the 9th column (<em>e.g.</em> transcript_id) and the term <strong>value</strong> for its associated string (e.g. “NM_334567”). The eight first columns of the GTF file (chrom/seqid, source, type, start, end, score, strand, frame) will be refered as <strong>basic attributes</strong>. In the example below, gene_id is the attribute and ‘G0001’ is the associated value.</p>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example| gtftk select_by_key -k feature -v gene| head -1
chr1	gtftk	gene	125	138	.	+	.	gene_id &quot;G0001&quot;;
</pre></div>
</div>
<hr class="docutils" />
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">About pygtftk</a><ul>
<li><a class="reference internal" href="#warning-about-supported-gtf-file-formats">warning about supported GTF file formats</a></li>
<li><a class="reference internal" href="#naming-conventions">Naming conventions</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">&lt;no title&gt;</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="installation.html"
                        title="next chapter">Installing pygtftk/gtftk</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/about.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installing pygtftk/gtftk"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="&lt;no title&gt;"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 0.9.6 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, F. Lopez, G. Charbonnier, A. Gonzalez, S. Spicuglia and D. Puthier.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.0.
    </div>
  </body>
</html>