

gtf_help_message = """
A GTF file containing genomic annotations (Ensembl recommended).
Gives counts of how may editing sites fall in what kind of features e.g. introns, exons.
If a bed file of editing islands is passed Adar1 and Adar2 counts will also be printed.

"""

min_site_coverage_message = """
Ignore sites with coverage less than N. May require an alignment file depending on the vcf file.

"""


def _shared_params(parser, coverage=True, min_site_coverage=True, gtf=True, editing_islands=True, names=True):
    """ This function stores basic argparse arguments used in almost every other function.
    """

    if min_site_coverage:
        parser.add_argument(
            "--min-site-coverage",
            type=int,
            default=1,
            help='Ignore sites with coverage less than N. May require an alignment file depending on the vcf file.'
        )

    if coverage:
        parser.add_argument(
            "--coverage",
            type=str,
            nargs='+',
            default=None,
            help='A coverage file generated by " samtools view -b {input} | genomeCoverageBed -ibam stdin -g > {output} " '
        )

    if gtf:
        parser.add_argument(
            "--gtf",
            type=str,
            help=gtf_help_message
        )

    if editing_islands:
        parser.add_argument(
            "--editing-islands",
            type=str,
            help='A bed file containing locations of editing islands.'
        )

    if names:
        parser.add_argument(
            "--names",
            type=str,
            help='Replace the file names in output with these names.'
        )

    parser.add_argument(
        "vcf",
        nargs='+',
        type=str,
        help='VCF formatted files containing editing sites.')


"""

    parser.add_argument(
        "--min-site-coverage",
        type=int,
        default=1,
        help='Ignore sites with coverage less than N. May require an alignment file depending on the vcf file.'
    )

    parser.add_argument(
        "--gtf",
        type=str,
        help='A GTF file containing genomic annotations (Ensembl recommended).' +
             'Gives counts of how may editing sites fall in what kind of features e.g. 3 prime UTRs, introns, exons.' +
             'If a bed file of editing islands is passed Adar1 and Adar2 counts will also be printed.'
    )

    parser.add_argument(
        "--editing-islands",
        type=str,
        help='A bed file containing locations of editing islands.'
    )

    parser.add_argument(
        "--names",
        type=str,
        help=''
    )

    parser.add_argument(
        "--coverage",
        type=str,
        nargs='+',
        default=None,
        help='A coverage file generated by " samtools view -b {input} | genomeCoverageBed -ibam stdin -g > {output} " '
    )

    parser.add_argument(
        "vcf_files",
        nargs='+',
        type=str,
        help='VCF formatted files containing editing sites.')

"""